development/secret-contracts/best-practices/data-handling.md

```diff
@@ -2,14 +2,82 @@
 
 ## Uint128
 
-### When To Use Them Instead Of u128
+`Uint128` is a data structure designed to work with usigned 128-bit integers.
 
-### Data Hygiene&#x20;
+If you are familiar with Rust, you might know that it has it's own native primitive - `u128`. `Uint128` differs from `u128` in that it is a string encoded integer, rather than the traditional little/big-endian.
 
-### How Uint128s Can Become A Security Risk&#x20;
+Simliarly, `cosmwasm-std` also has `Uint64` and `Uint256` and all of the following applies there as well.
+
+### When To Use Uint128 Instead Of u128
+
+`Uint128` is a thin wrapper around `u128` that is using strings for JSON encoding/decoding, such that the full `u128` range can be used for clients that convert JSON numbers to floats, like JavaScript and jq ([source](https://docs.cosmwasm.com/docs/1.0/smart-contracts/math/#uint128)).
+
+### Entrypoint Messages
+
+If you are familiar with Messages, you already know that most of the time we will use [serde](https://serde.rs/) to deserialize them from JSON (if not, you should read on [contract entrypoints](https://docs.cosmwasm.com/docs/1.0/actor-model/actor-in-blokchain/#entry-points) and the concept of [Messages](https://docs.cosmwasm.com/docs/1.0/smart-contracts/message/message)). Output will often be serizlized in the same way.
+
+In general, JSON implementations usually accept `[-(2^53)+1,(2^53)-1]` as an acceptable range for numbers. That's why we'll prefer to use `Uint128` in entrypoint messages, for example:
```

Comment:
maybe strengthen connection between [range in json] to [reason to use Uint128]. e.g. "if we need more than (that/53 bits)"

Reply:
Can you give a full example? Not sure where to put it